File: leetcode/0145_binary_tree_postorder_traversal.cc
#include <vector>
#include <stack>
using namespace std;
/*
    145 二叉树的后序遍历
        给定一个二叉树，返回它的 后序 遍历

*/

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

// 非递归版本
class Solution {
public:
    vector<int> postorderTraversal(TreeNode* root) {
        stack<TreeNode*> st;
        if (!root) { 
            return {}; 
        }
        vector<int> ret;
        st.push(root);
        while (!st.empty()) {
            TreeNode* node = st.top();
            st.pop();
            ret.push_back(node->val);
            if (node->left) {
                st.push(node->left);
            }
            if (node->right) {
                st.push(node->right);
            }
        }
        reverse(ret.begin(), ret.end());
        return ret;
    }
};

// 递归版本
class Solution1 {
public:
	vector<int> postorderTraversal(TreeNode* root) {
        if (!root) {
            return {};
        }
        
        vector<int> ret;
        postorderTraversal(root, ret);
        return ret;
	}

	void postorderTraversal(TreeNode* node, vector<int>& ret) {
        if (!node) {
            return;
        }

        postorderTraversal(node->left, ret);
        postorderTraversal(node->right, ret);
        ret.push_back(node->val);
	}
};

File: leetcode/0145_binary_tree_postorder_traversal.java
import java.util.ArrayList;
import java.util.List;
import java.util.Stack;

class Solution {
    public List<Integer> postorderTraversal(TreeNode root) {
        List<Integer> result = new ArrayList<>();
        if (root != null) {
            Stack<TreeNode> stack = new Stack<>();
            stack.push(root);
            Stack<TreeNode> ret_st = new Stack<>();
            while (!stack.empty()) {
                TreeNode node = stack.pop();
                ret_st.push(node);
                if (node.left != null) {
                    stack.push(node.left);
                }
                if (node.right != null) {
                    stack.push(node.right);
                }
            }
            while (!ret_st.empty()) {
                TreeNode node = ret_st.pop();
                result.add(node.val);
            }
        }

        return result;
    }

    public class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;
        TreeNode() {}
        TreeNode(int val) { this.val = val; }
        TreeNode(int val, TreeNode left, TreeNode right) {
            this.val = val;
            this.left = left;
            this.right = right;
        }
    }
}